text/0000-ptr-meta.md

```diff
@@ -0,0 +1,320 @@
+        std::alloc::dealloc(raw as *mut u8, vtable.layout());
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+(Parts of this assumes that [trait aliases] are implemented.
+If they are not implemented by the time this RFC is,
+the `Thin` alias can be replaced by its definition.)
+
+The APIs whose full definition is found below
+are added to `core::ptr` and re-exported in `std::ptr`:
+
+* A `Pointee` trait,
+  implemented automatically for all types
+  (similar to how `Sized` and `Unsize` are implemented automatically).
+* A `Thin` [trait alias].
+  If this RFC is implemented before type aliases are,
+  uses of `Thin` should be replaced with its definition.
+* A `metadata` free function
+* A `from_raw_parts` constructor for each of `*const T` and `*mut T`
+
+The bounds on `null()` and `null_mut()` function in that same module
+as well as the `NonNull::dangling` constructor
+are changed from (implicit) `T: Sized` to `T: ?Sized + Thin`.
+This enables using those functions with [extern types].
+
+For the purpose of pointer casts being allowed by the `as` operator,
+a pointer to `T` is considered to be thin if `T: Thin` instead of `T: Sized`.
+This similarly includes extern types.
+
+`std::raw::TraitObject` and `std::raw` are deprecated and eventually removed.
+
+[trait alias]: https://github.com/rust-lang/rust/issues/41517
+[extern types]: https://github.com/rust-lang/rust/issues/43467
+
+```rust
+/// This trait is automatically implement for every type.
+///
+/// Raw pointer types and referenece types in Rust can be thought of as made of two parts:
+/// a data pointer that contains the memory address of the value, and some metadata.
+///
+/// For statically-sized types that implement the `Sized` traits,
+/// pointers are said to be “thin”, metadata is zero-size, and metadata’s type is `()`.
+///
+/// Pointers to [dynamically-sized types][dst] are said to be “fat”
+/// and have non-zero-size metadata:
+///
+/// * For structs whose last field is a DST, metadata is the metadata for the last field
+/// * For the `str` type, metadata is the length in bytes as `usize`
+/// * For slice types like `[T]`, metadata is the length in items as `usize`
+/// * For trait objects like `dyn SomeTrait`, metadata is [`&'static VTable`][VTable]
+///
+/// In the future, the Rust language may gain new kinds of types
+/// that have different pointer metadata.
+///
+/// Pointer metadata can be extracted from a pointer or reference with the [`metadata`] function.
+/// The data pointer can be extracted by casting a (fat) pointer
+/// to a (thin) pointer to a `Sized` type the `as` operator,
+/// for example `(x: &dyn SomeTrait) as *const SomeTrait as *const ()`.
+///
+/// [dst]: https://doc.rust-lang.org/nomicon/exotic-sizes.html#dynamically-sized-types-dsts
+#[lang = "pointee"]
+pub trait Pointee {
+    /// The type for metadata in pointers and references to `Self`.
+    type Metadata;
```

Comment:
For future compatibility I strongly suggest we bound the `Metadata` type as

```suggestion
    type Metadata: Copy + Send + Sync + Ord + Hash + 'static;
```

The metadata of both `[T]` and `dyn Tr` satisfies these bounds (being `usize` and `&'static VTable` respectively)

Rationale is in https://github.com/kennytm/rfcs/blob/dyn-type/text/0000-dyn-type/31-Custom-DST-from-scratch.md#object-trait.

Reply:
I think that as far as future-compat is concerned it is not urgent to have these bounds as part of the associated type’s definition. As long as custom DSTs do not exist, all impls of `Pointee` are compiler-generated and happen to satisfy those bounds already.

However making those bounds explicit could help generic code that works with any `?Sized` type and manipulates metadata. I’ll add them.